pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctAggregationFunction.java

```diff
@@ -0,0 +1,165 @@
+
+import com.google.common.base.Preconditions;
+import java.util.Arrays;
+import java.util.Iterator;
+import javax.annotation.Nonnull;
+import org.apache.pinot.common.data.FieldSpec;
+import org.apache.pinot.common.function.AggregationFunctionType;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.common.BlockValSet;
+import org.apache.pinot.core.data.table.Key;
+import org.apache.pinot.core.operator.transform.TransformBlockDataFetcher;
+import org.apache.pinot.core.operator.transform.TransformResultMetadata;
+import org.apache.pinot.core.query.aggregation.AggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.DistinctTable;
+import org.apache.pinot.core.query.aggregation.ObjectAggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.groupby.GroupByResultHolder;
+import org.apache.pinot.core.segment.index.readers.Dictionary;
+import org.apache.pinot.pql.parsers.pql2.ast.FunctionCallAstNode;
+
+
+/**
+ * DISTINCT clause in SQL is implemented as function in the
+ * execution engine of Pinot.
+ */
+public class DistinctAggregationFunction implements AggregationFunction<DistinctTable, Comparable> {
+
+  private final DistinctTable _distinctTable;
+  private final String[] columnNames;
+  private FieldSpec.DataType[] _dataTypes;
+  private boolean init = false;
+
+  DistinctAggregationFunction(String multiColumnExpression, int limit) {
+    _distinctTable = new DistinctTable(limit);
+    columnNames = multiColumnExpression.split(FunctionCallAstNode.DISTINCT_MULTI_COLUMN_SEPARATOR);
+    _dataTypes = new FieldSpec.DataType[columnNames.length];
+  }
+
+  @Nonnull
+  @Override
+  public AggregationFunctionType getType() {
+    return AggregationFunctionType.DISTINCT;
+  }
+
+  @Nonnull
+  @Override
+  public String getColumnName(@Nonnull String column) {
+    return AggregationFunctionType.DISTINCT.getName() + "_" + column;
+  }
+
+  @Nonnull
+  @Override
+  public DataSchema.ColumnDataType getIntermediateResultColumnType() {
+    return DataSchema.ColumnDataType.OBJECT;
+  }
+
+  @Override
+  public void accept(@Nonnull AggregationFunctionVisitorBase visitor) {
+    visitor.visit(this);
+  }
+
+  @Nonnull
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return new ObjectAggregationResultHolder();
+  }
+
+  @Override
+  public void aggregate(int records, AggregationResultHolder aggregationResultHolder,
+      @Nonnull BlockValSet... blockValSets) {
+    Preconditions.checkArgument(blockValSets.length == columnNames.length, "Error invalid number of block value sets");
+
+    if (!init) {
+      for (int i = 0; i < blockValSets.length; i++) {
+        _dataTypes[i] = blockValSets[i].getValueType();
+      }
+      _distinctTable.setProjectedColumnNames(columnNames);
+      _distinctTable.setProjectedColumnTypes(_dataTypes);
+      init = true;
+    }
+
+    // TODO: Follow up PR will make few changes to start using DictionaryBasedAggregationOperator
```

Comment:
Without filter and on single column

Reply:
Exactly -- that's what we decided in the design doc